File: frontend/public/python/league_financials.py
"""
league_financials.py
Central financial rules and 6.5% annual league inflation helpers.

Existing signed contracts should stay fixed. These helpers update only league
rules used for future contracts, cap holds, exceptions, rookie deals, and UI
cap values.
"""
from __future__ import annotations

import copy
import math
import datetime as _dt
from typing import Any, Dict, Optional

LEAGUE_FINANCIALS_VERSION = "2026-06-14_league_inflation_v1"
DEFAULT_ANNUAL_INFLATION_RATE = 0.065
DEFAULT_BASE_SEASON_YEAR = 2026

DEFAULT_BASE_FINANCIAL_RULES: Dict[str, float] = {
    "salaryCap": 154_647_000,
    "luxuryTaxLine": 187_895_000,
    "firstApron": 195_945_000,
    "secondApron": 207_824_000,
    "hardCap": 207_824_000,

    "minimumSalary": 1_200_000,
    "minimumException": 1_500_000,
    "veteranMinimum": 1_500_000,
    "twoWaySalary": 580_000,

    "maxSalary": 54_000_000,
    "roomException": 8_781_000,
    "nonTaxpayerMLE": 14_104_000,
    "midLevelException": 14_104_000,
    "taxpayerMLE": 5_685_000,

    "rookiePick1Salary": 11_800_000,
    "rookieFirstRoundDecline": 315_000,
    "rookieFirstRoundFloor": 2_400_000,
    "rookieSecondRoundBase": 2_250_000,
    "rookieSecondRoundDecline": 28_000,
    "rookieSecondRoundFloor": 1_250_000,
}


def _num(value: Any, fallback: float = 0.0) -> float:
    try:
        if value is None:
            return float(fallback)
        return float(value)
    except Exception:
        return float(fallback)


def _season_year(value: Any, fallback: int = DEFAULT_BASE_SEASON_YEAR) -> int:
    try:
        y = int(float(value))
        if 2020 <= y <= 2100:
            return y
    except Exception:
        pass
    return int(fallback)


def round_money(value: float, nearest: int = 1_000) -> int:
    base = max(1, int(nearest or 1))
    return int(base * round(float(value or 0) / base))


def get_league_season_year(league_data: Optional[Dict[str, Any]]) -> int:
    league_data = league_data or {}
    return _season_year(
        league_data.get("seasonYear")
        or league_data.get("currentSeasonYear")
        or league_data.get("seasonStartYear"),
        DEFAULT_BASE_SEASON_YEAR,
    )


def get_current_financial_season_year(league_data: Optional[Dict[str, Any]]) -> int:
    league_data = league_data or {}
    financials = league_data.get("financials") if isinstance(league_data.get("financials"), dict) else {}
    return _season_year(
        league_data.get("currentFinancialSeasonYear")
        or financials.get("currentSeasonYear")
        or financials.get("currentFinancialSeasonYear")
        or financials.get("appliedThroughSeasonYear")
        or financials.get("appliedInflationThroughSeason")
        or get_league_season_year(league_data),
        get_league_season_year(league_data),
    )


def _build_base_rules_from_league(league_data: Optional[Dict[str, Any]]) -> Dict[str, float]:
    league_data = league_data or {}
    financials = league_data.get("financials") if isinstance(league_data.get("financials"), dict) else {}
    existing_base = financials.get("baseRules") if isinstance(financials.get("baseRules"), dict) else {}
    out = dict(DEFAULT_BASE_FINANCIAL_RULES)
    out.update(existing_base)

    out["salaryCap"] = _num(existing_base.get("salaryCap") or league_data.get("salaryCap") or league_data.get("capLimit"), out["salaryCap"])
    out["luxuryTaxLine"] = _num(existing_base.get("luxuryTaxLine") or league_data.get("luxuryTaxLine") or league_data.get("taxLine"), out["luxuryTaxLine"])
    out["firstApron"] = _num(existing_base.get("firstApron") or league_data.get("firstApron") or league_data.get("apron1"), out["firstApron"])
    out["secondApron"] = _num(existing_base.get("secondApron") or league_data.get("secondApron") or league_data.get("apron2"), out["secondApron"])
    out["hardCap"] = _num(existing_base.get("hardCap") or league_data.get("hardCap") or league_data.get("hardCapLimit") or league_data.get("secondApron") or league_data.get("apron2"), out["hardCap"])

    out["minimumSalary"] = _num(existing_base.get("minimumSalary") or league_data.get("minimumSalary"), out["minimumSalary"])
    out["minimumException"] = _num(existing_base.get("minimumException") or league_data.get("minimumException"), out["minimumException"])
    out["veteranMinimum"] = _num(existing_base.get("veteranMinimum") or league_data.get("veteranMinimum") or league_data.get("minimumException"), out["veteranMinimum"])
    out["twoWaySalary"] = _num(existing_base.get("twoWaySalary") or league_data.get("twoWaySalary"), out["twoWaySalary"])

    out["maxSalary"] = _num(existing_base.get("maxSalary") or league_data.get("maxSalary") or league_data.get("maxContract") or league_data.get("maxContractAmount"), out["maxSalary"])
    out["roomException"] = _num(existing_base.get("roomException") or league_data.get("roomException") or league_data.get("roomExceptionAmount"), out["roomException"])
    out["nonTaxpayerMLE"] = _num(existing_base.get("nonTaxpayerMLE") or league_data.get("nonTaxpayerMLE") or league_data.get("nonTaxpayerMidLevelException") or league_data.get("midLevelException"), out["nonTaxpayerMLE"])
    out["midLevelException"] = _num(existing_base.get("midLevelException") or league_data.get("midLevelException") or league_data.get("nonTaxpayerMLE") or league_data.get("nonTaxpayerMidLevelException"), out["midLevelException"])
    out["taxpayerMLE"] = _num(existing_base.get("taxpayerMLE") or league_data.get("taxpayerMLE") or league_data.get("taxpayerMidLevelException"), out["taxpayerMLE"])

    for key in [
        "rookiePick1Salary", "rookieFirstRoundDecline", "rookieFirstRoundFloor",
        "rookieSecondRoundBase", "rookieSecondRoundDecline", "rookieSecondRoundFloor",
    ]:
        out[key] = _num(existing_base.get(key), DEFAULT_BASE_FINANCIAL_RULES[key])

    return out


def _inflation_index(base_season_year: int, season_year: int, annual_rate: float) -> float:
    years = max(0, _season_year(season_year) - _season_year(base_season_year))
    return math.pow(1.0 + float(annual_rate), years)


def get_financial_rules(league_data: Optional[Dict[str, Any]], season_year: Optional[int] = None) -> Dict[str, Any]:
    league_data = league_data or {}
    financials = league_data.get("financials") if isinstance(league_data.get("financials"), dict) else {}
    base_season_year = _season_year(financials.get("baseSeasonYear") or get_league_season_year(league_data))
    current_year = _season_year(season_year or get_current_financial_season_year(league_data), base_season_year)
    annual_rate = _num(financials.get("annualInflationRate"), DEFAULT_ANNUAL_INFLATION_RATE)
    base_rules = _build_base_rules_from_league(league_data)
    index = _inflation_index(base_season_year, current_year, annual_rate)

    def scaled(key: str, nearest: int = 1_000) -> int:
        return round_money(_num(base_rules.get(key), DEFAULT_BASE_FINANCIAL_RULES.get(key, 0)) * index, nearest)

    rules = {
        "version": LEAGUE_FINANCIALS_VERSION,
        "baseSeasonYear": base_season_year,
        "seasonYear": current_year,
        "currentFinancialSeasonYear": current_year,
        "annualInflationRate": annual_rate,
        "inflationIndex": index,

        "salaryCap": scaled("salaryCap"),
        "luxuryTaxLine": scaled("luxuryTaxLine"),
        "firstApron": scaled("firstApron"),
        "secondApron": scaled("secondApron"),
        "hardCap": scaled("hardCap"),

        "minimumSalary": scaled("minimumSalary"),
        "minimumException": scaled("minimumException"),
        "veteranMinimum": scaled("veteranMinimum"),
        "twoWaySalary": scaled("twoWaySalary"),

        "maxSalary": scaled("maxSalary"),
        "roomException": scaled("roomException"),
        "nonTaxpayerMLE": scaled("nonTaxpayerMLE"),
        "midLevelException": scaled("nonTaxpayerMLE"),
        "taxpayerMLE": scaled("taxpayerMLE"),
    }
    rules["capLimit"] = rules["salaryCap"]
    rules["taxLine"] = rules["luxuryTaxLine"]
    rules["apron1"] = rules["firstApron"]
    rules["apron2"] = rules["secondApron"]
    rules["hardCapLimit"] = rules["hardCap"]
    rules["maxContract"] = rules["maxSalary"]
    rules["maxContractAmount"] = rules["maxSalary"]
    rules["roomExceptionAmount"] = rules["roomException"]
    rules["nonTaxpayerMidLevelException"] = rules["nonTaxpayerMLE"]
    rules["taxpayerMidLevelException"] = rules["taxpayerMLE"]
    return rules


def normalize_financial_aliases(league_data: Dict[str, Any], rules: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    if not isinstance(league_data, dict):
        return league_data
    rules = rules or get_financial_rules(league_data)
    league_data["currentFinancialSeasonYear"] = rules["seasonYear"]
    league_data["salaryCap"] = rules["salaryCap"]
    league_data["capLimit"] = rules["salaryCap"]
    league_data["luxuryTaxLine"] = rules["luxuryTaxLine"]
    league_data["taxLine"] = rules["luxuryTaxLine"]
    league_data["firstApron"] = rules["firstApron"]
    league_data["apron1"] = rules["firstApron"]
    league_data["secondApron"] = rules["secondApron"]
    league_data["apron2"] = rules["secondApron"]
    league_data["hardCap"] = rules["hardCap"]
    league_data["hardCapLimit"] = rules["hardCap"]
    league_data["minimumSalary"] = rules["minimumSalary"]
    league_data["minimumException"] = rules["minimumException"]
    league_data["veteranMinimum"] = rules["veteranMinimum"]
    league_data["twoWaySalary"] = rules["twoWaySalary"]
    league_data["maxSalary"] = rules["maxSalary"]
    league_data["maxContract"] = rules["maxSalary"]
    league_data["maxContractAmount"] = rules["maxSalary"]
    league_data["roomException"] = rules["roomException"]
    league_data["roomExceptionAmount"] = rules["roomException"]
    league_data["midLevelException"] = rules["midLevelException"]
    league_data["nonTaxpayerMLE"] = rules["nonTaxpayerMLE"]
    league_data["nonTaxpayerMidLevelException"] = rules["nonTaxpayerMLE"]
    league_data["taxpayerMLE"] = rules["taxpayerMLE"]
    league_data["taxpayerMidLevelException"] = rules["taxpayerMLE"]
    return league_data


def ensure_league_financials(league_data: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    if not isinstance(league_data, dict):
        return league_data

    financials = league_data.get("financials") if isinstance(league_data.get("financials"), dict) else {}
    base_season_year = _season_year(financials.get("baseSeasonYear") or get_league_season_year(league_data))
    current_year = _season_year(
        financials.get("currentSeasonYear")
        or financials.get("currentFinancialSeasonYear")
        or league_data.get("currentFinancialSeasonYear")
        or financials.get("appliedThroughSeasonYear")
        or financials.get("appliedInflationThroughSeason")
        or get_league_season_year(league_data),
        base_season_year,
    )
    annual_rate = _num(financials.get("annualInflationRate"), DEFAULT_ANNUAL_INFLATION_RATE)
    base_rules = _build_base_rules_from_league(league_data)
    applied_year = _season_year(
        financials.get("appliedThroughSeasonYear")
        or financials.get("appliedInflationThroughSeason")
        or current_year,
        current_year,
    )

    next_financials = dict(financials)
    next_financials.update({
        "version": LEAGUE_FINANCIALS_VERSION,
        "baseSeasonYear": base_season_year,
        "annualInflationRate": annual_rate,
        "baseRules": base_rules,
        "currentSeasonYear": current_year,
        "currentFinancialSeasonYear": current_year,
        "appliedThroughSeasonYear": applied_year,
    })
    history = next_financials.get("history") if isinstance(next_financials.get("history"), dict) else {}
    next_financials["history"] = history
    league_data["financials"] = next_financials

    rules = get_financial_rules(league_data, current_year)
    history[str(current_year)] = {
        **(history.get(str(current_year)) if isinstance(history.get(str(current_year)), dict) else {}),
        "seasonYear": current_year,
        "inflationIndex": rules["inflationIndex"],
        "salaryCap": rules["salaryCap"],
        "luxuryTaxLine": rules["luxuryTaxLine"],
        "firstApron": rules["firstApron"],
        "secondApron": rules["secondApron"],
        "hardCap": rules["hardCap"],
        "minimumSalary": rules["minimumSalary"],
        "minimumException": rules["minimumException"],
        "maxSalary": rules["maxSalary"],
        "midLevelException": rules["midLevelException"],
        "taxpayerMLE": rules["taxpayerMLE"],
        "roomException": rules["roomException"],
        "twoWaySalary": rules["twoWaySalary"],
    }
    return normalize_financial_aliases(league_data, rules)


def apply_league_inflation_for_offseason(league_data: Dict[str, Any], target_season_year: Optional[int] = None) -> Dict[str, Any]:
    league_data = ensure_league_financials(copy.deepcopy(league_data or {}))
    target_year = _season_year(target_season_year or get_league_season_year(league_data) + 1)
    financials = league_data.get("financials", {})
    applied_year = _season_year(
        financials.get("appliedThroughSeasonYear") or financials.get("appliedInflationThroughSeason") or financials.get("currentSeasonYear"),
        get_league_season_year(league_data),
    )

    if applied_year < target_year:
        financials["currentSeasonYear"] = target_year
        financials["currentFinancialSeasonYear"] = target_year
        financials["appliedThroughSeasonYear"] = target_year
        financials["appliedInflationThroughSeason"] = target_year
        financials["lastAppliedAt"] = _dt.datetime.utcnow().isoformat() + "Z"
        league_data["currentFinancialSeasonYear"] = target_year

    rules = get_financial_rules(league_data, target_year)
    history = financials.get("history") if isinstance(financials.get("history"), dict) else {}
    financials["history"] = history
    history[str(target_year)] = {
        "seasonYear": target_year,
        "inflationIndex": rules["inflationIndex"],
        "salaryCap": rules["salaryCap"],
        "luxuryTaxLine": rules["luxuryTaxLine"],
        "firstApron": rules["firstApron"],
        "secondApron": rules["secondApron"],
        "hardCap": rules["hardCap"],
        "minimumSalary": rules["minimumSalary"],
        "minimumException": rules["minimumException"],
        "veteranMinimum": rules["veteranMinimum"],
        "maxSalary": rules["maxSalary"],
        "midLevelException": rules["midLevelException"],
        "taxpayerMLE": rules["taxpayerMLE"],
        "roomException": rules["roomException"],
        "twoWaySalary": rules["twoWaySalary"],
        "appliedAt": financials.get("lastAppliedAt"),
    }
    return normalize_financial_aliases(league_data, rules)


def get_rookie_salary_for_pick(league_data: Optional[Dict[str, Any]], round_num: int, pick_num: int, season_year: Optional[int] = None) -> int:
    league_data = ensure_league_financials(league_data or {})
    rules = get_financial_rules(league_data, season_year or get_current_financial_season_year(league_data))
    base_rules = league_data.get("financials", {}).get("baseRules", DEFAULT_BASE_FINANCIAL_RULES)
    index = _num(rules.get("inflationIndex"), 1.0)

    if int(round_num or 1) == 1:
        base_salary = max(
            _num(base_rules.get("rookieFirstRoundFloor"), DEFAULT_BASE_FINANCIAL_RULES["rookieFirstRoundFloor"]),
            _num(base_rules.get("rookiePick1Salary"), DEFAULT_BASE_FINANCIAL_RULES["rookiePick1Salary"]) -
            (max(1, int(pick_num or 1)) - 1) * _num(base_rules.get("rookieFirstRoundDecline"), DEFAULT_BASE_FINANCIAL_RULES["rookieFirstRoundDecline"]),
        )
    else:
        pick_in_round = max(1, int(pick_num or 31) - 30)
        base_salary = max(
            _num(base_rules.get("rookieSecondRoundFloor"), DEFAULT_BASE_FINANCIAL_RULES["rookieSecondRoundFloor"]),
            _num(base_rules.get("rookieSecondRoundBase"), DEFAULT_BASE_FINANCIAL_RULES["rookieSecondRoundBase"]) -
            (pick_in_round - 1) * _num(base_rules.get("rookieSecondRoundDecline"), DEFAULT_BASE_FINANCIAL_RULES["rookieSecondRoundDecline"]),
        )

    return round_money(base_salary * index, 1_000)

File: _patch_backups/season_context_2026_27_20260725-212723/frontend/src/utils/leagueFinancials.js
// src/utils/leagueFinancials.js
// Central league economy helpers for salary-cap inflation.
// Existing signed contracts should remain fixed. These helpers only set the
// financial rules used for future offers, rookie deals, cap holds, exceptions,
// and UI displays.

export const LEAGUE_FINANCIALS_VERSION = "2026-06-14_league_inflation_v1";
export const DEFAULT_ANNUAL_INFLATION_RATE = 0.065;
export const DEFAULT_BASE_SEASON_YEAR = 2026;

export const DEFAULT_BASE_FINANCIAL_RULES = Object.freeze({
  salaryCap: 154_647_000,
  luxuryTaxLine: 187_895_000,
  firstApron: 195_945_000,
  secondApron: 207_824_000,
  hardCap: 207_824_000,

  minimumSalary: 1_200_000,
  minimumException: 1_500_000,
  veteranMinimum: 1_500_000,
  twoWaySalary: 580_000,

  maxSalary: 54_000_000,
  roomException: 8_781_000,
  nonTaxpayerMLE: 14_104_000,
  midLevelException: 14_104_000,
  taxpayerMLE: 5_685_000,

  rookiePick1Salary: 11_800_000,
  rookieFirstRoundDecline: 315_000,
  rookieFirstRoundFloor: 2_400_000,
  rookieSecondRoundBase: 2_250_000,
  rookieSecondRoundDecline: 28_000,
  rookieSecondRoundFloor: 1_250_000,
});

function safeNumber(value, fallback = 0) {
  const n = Number(value);
  return Number.isFinite(n) ? n : fallback;
}

export function roundMoney(value, nearest = 1_000) {
  const base = Math.max(1, Number(nearest) || 1);
  return Math.round((Number(value) || 0) / base) * base;
}

function validSeasonYear(value, fallback = DEFAULT_BASE_SEASON_YEAR) {
  const y = Number(value);
  if (Number.isFinite(y) && y >= 2020 && y <= 2100) return Math.round(y);
  return fallback;
}

export function getLeagueSeasonYear(leagueData = {}) {
  return validSeasonYear(
    leagueData?.seasonYear ?? leagueData?.currentSeasonYear ?? leagueData?.seasonStartYear,
    DEFAULT_BASE_SEASON_YEAR
  );
}

export function getCurrentFinancialSeasonYear(leagueData = {}) {
  const financials = leagueData?.financials || {};
  return validSeasonYear(
    leagueData?.currentFinancialSeasonYear ??
      financials?.currentSeasonYear ??
      financials?.currentFinancialSeasonYear ??
      financials?.appliedThroughSeasonYear ??
      financials?.appliedInflationThroughSeason ??
      getLeagueSeasonYear(leagueData),
    getLeagueSeasonYear(leagueData)
  );
}

function buildBaseRulesFromLeague(leagueData = {}) {
  const top = leagueData || {};
  const financials = top.financials || {};
  const existingBase = financials.baseRules && typeof financials.baseRules === "object"
    ? financials.baseRules
    : {};

  return {
    ...DEFAULT_BASE_FINANCIAL_RULES,
    ...existingBase,

    salaryCap: safeNumber(existingBase.salaryCap ?? top.salaryCap ?? top.capLimit, DEFAULT_BASE_FINANCIAL_RULES.salaryCap),
    luxuryTaxLine: safeNumber(existingBase.luxuryTaxLine ?? top.luxuryTaxLine ?? top.taxLine, DEFAULT_BASE_FINANCIAL_RULES.luxuryTaxLine),
    firstApron: safeNumber(existingBase.firstApron ?? top.firstApron ?? top.apron1, DEFAULT_BASE_FINANCIAL_RULES.firstApron),
    secondApron: safeNumber(existingBase.secondApron ?? top.secondApron ?? top.apron2, DEFAULT_BASE_FINANCIAL_RULES.secondApron),
    hardCap: safeNumber(existingBase.hardCap ?? top.hardCap ?? top.hardCapLimit ?? top.secondApron ?? top.apron2, DEFAULT_BASE_FINANCIAL_RULES.hardCap),

    minimumSalary: safeNumber(existingBase.minimumSalary ?? top.minimumSalary ?? top.minSalary, DEFAULT_BASE_FINANCIAL_RULES.minimumSalary),
    minimumException: safeNumber(existingBase.minimumException ?? top.minimumException, DEFAULT_BASE_FINANCIAL_RULES.minimumException),
    veteranMinimum: safeNumber(existingBase.veteranMinimum ?? top.veteranMinimum ?? top.minimumException, DEFAULT_BASE_FINANCIAL_RULES.veteranMinimum),
    twoWaySalary: safeNumber(existingBase.twoWaySalary ?? top.twoWaySalary, DEFAULT_BASE_FINANCIAL_RULES.twoWaySalary),

    maxSalary: safeNumber(existingBase.maxSalary ?? top.maxSalary ?? top.maxContract ?? top.maxContractAmount, DEFAULT_BASE_FINANCIAL_RULES.maxSalary),
    roomException: safeNumber(existingBase.roomException ?? top.roomException ?? top.roomExceptionAmount, DEFAULT_BASE_FINANCIAL_RULES.roomException),
    nonTaxpayerMLE: safeNumber(existingBase.nonTaxpayerMLE ?? top.nonTaxpayerMLE ?? top.nonTaxpayerMidLevelException ?? top.midLevelException, DEFAULT_BASE_FINANCIAL_RULES.nonTaxpayerMLE),
    midLevelException: safeNumber(existingBase.midLevelException ?? top.midLevelException ?? top.nonTaxpayerMLE ?? top.nonTaxpayerMidLevelException, DEFAULT_BASE_FINANCIAL_RULES.midLevelException),
    taxpayerMLE: safeNumber(existingBase.taxpayerMLE ?? top.taxpayerMLE ?? top.taxpayerMidLevelException, DEFAULT_BASE_FINANCIAL_RULES.taxpayerMLE),

    rookiePick1Salary: safeNumber(existingBase.rookiePick1Salary, DEFAULT_BASE_FINANCIAL_RULES.rookiePick1Salary),
    rookieFirstRoundDecline: safeNumber(existingBase.rookieFirstRoundDecline, DEFAULT_BASE_FINANCIAL_RULES.rookieFirstRoundDecline),
    rookieFirstRoundFloor: safeNumber(existingBase.rookieFirstRoundFloor, DEFAULT_BASE_FINANCIAL_RULES.rookieFirstRoundFloor),
    rookieSecondRoundBase: safeNumber(existingBase.rookieSecondRoundBase, DEFAULT_BASE_FINANCIAL_RULES.rookieSecondRoundBase),
    rookieSecondRoundDecline: safeNumber(existingBase.rookieSecondRoundDecline, DEFAULT_BASE_FINANCIAL_RULES.rookieSecondRoundDecline),
    rookieSecondRoundFloor: safeNumber(existingBase.rookieSecondRoundFloor, DEFAULT_BASE_FINANCIAL_RULES.rookieSecondRoundFloor),
  };
}

function calculateInflationIndex(baseSeasonYear, seasonYear, annualRate) {
  const years = Math.max(0, validSeasonYear(seasonYear) - validSeasonYear(baseSeasonYear));
  return Math.pow(1 + safeNumber(annualRate, DEFAULT_ANNUAL_INFLATION_RATE), years);
}

export function getLeagueFinancialRules(leagueData = {}, seasonYear = null) {
  const financials = leagueData?.financials || {};
  const baseSeasonYear = validSeasonYear(financials.baseSeasonYear ?? getLeagueSeasonYear(leagueData));
  const currentYear = validSeasonYear(seasonYear ?? getCurrentFinancialSeasonYear(leagueData), baseSeasonYear);
  const annualInflationRate = safeNumber(financials.annualInflationRate, DEFAULT_ANNUAL_INFLATION_RATE);
  const baseRules = buildBaseRulesFromLeague(leagueData);
  const inflationIndex = calculateInflationIndex(baseSeasonYear, currentYear, annualInflationRate);

  const scaled = (key, nearest = 1_000) => roundMoney(safeNumber(baseRules[key], DEFAULT_BASE_FINANCIAL_RULES[key]) * inflationIndex, nearest);

  const rules = {
    version: LEAGUE_FINANCIALS_VERSION,
    baseSeasonYear,
    seasonYear: currentYear,
    currentFinancialSeasonYear: currentYear,
    annualInflationRate,
    inflationIndex,

    salaryCap: scaled("salaryCap"),
    luxuryTaxLine: scaled("luxuryTaxLine"),
    firstApron: scaled("firstApron"),
    secondApron: scaled("secondApron"),
    hardCap: scaled("hardCap"),

    minimumSalary: scaled("minimumSalary"),
    minimumException: scaled("minimumException"),
    veteranMinimum: scaled("veteranMinimum"),
    twoWaySalary: scaled("twoWaySalary"),

    maxSalary: scaled("maxSalary"),
    maxContract: scaled("maxSalary"),
    maxContractAmount: scaled("maxSalary"),
    roomException: scaled("roomException"),
    roomExceptionAmount: scaled("roomException"),
    nonTaxpayerMLE: scaled("nonTaxpayerMLE"),
    nonTaxpayerMidLevelException: scaled("nonTaxpayerMLE"),
    midLevelException: scaled("nonTaxpayerMLE"),
    taxpayerMLE: scaled("taxpayerMLE"),
    taxpayerMidLevelException: scaled("taxpayerMLE"),
  };

  rules.capLimit = rules.salaryCap;
  rules.taxLine = rules.luxuryTaxLine;
  rules.apron1 = rules.firstApron;
  rules.apron2 = rules.secondApron;
  rules.hardCapLimit = rules.hardCap;

  return rules;
}

export function normalizeLeagueFinancialAliases(leagueData = {}, rules = null) {
  if (!leagueData || typeof leagueData !== "object") return leagueData;
  const resolvedRules = rules || getLeagueFinancialRules(leagueData);
  return {
    ...leagueData,
    currentFinancialSeasonYear: resolvedRules.seasonYear,

    salaryCap: resolvedRules.salaryCap,
    capLimit: resolvedRules.salaryCap,
    luxuryTaxLine: resolvedRules.luxuryTaxLine,
    taxLine: resolvedRules.luxuryTaxLine,
    firstApron: resolvedRules.firstApron,
    apron1: resolvedRules.firstApron,
    secondApron: resolvedRules.secondApron,
    apron2: resolvedRules.secondApron,
    hardCap: resolvedRules.hardCap,
    hardCapLimit: resolvedRules.hardCap,

    minimumSalary: resolvedRules.minimumSalary,
    minimumException: resolvedRules.minimumException,
    veteranMinimum: resolvedRules.veteranMinimum,
    twoWaySalary: resolvedRules.twoWaySalary,
    maxSalary: resolvedRules.maxSalary,
    maxContract: resolvedRules.maxSalary,
    maxContractAmount: resolvedRules.maxSalary,

    roomException: resolvedRules.roomException,
    roomExceptionAmount: resolvedRules.roomException,
    midLevelException: resolvedRules.midLevelException,
    nonTaxpayerMLE: resolvedRules.nonTaxpayerMLE,
    nonTaxpayerMidLevelException: resolvedRules.nonTaxpayerMidLevelException,
    taxpayerMLE: resolvedRules.taxpayerMLE,
    taxpayerMidLevelException: resolvedRules.taxpayerMLE,
  };
}

export function ensureLeagueFinancials(leagueData = {}, options = {}) {
  if (!leagueData || typeof leagueData !== "object") return leagueData;

  const existing = leagueData.financials && typeof leagueData.financials === "object"
    ? leagueData.financials
    : {};

  const baseSeasonYear = validSeasonYear(
    existing.baseSeasonYear ?? options.baseSeasonYear ?? getLeagueSeasonYear(leagueData),
    DEFAULT_BASE_SEASON_YEAR
  );
  const currentFinancialSeasonYear = validSeasonYear(
    existing.currentSeasonYear ??
      existing.currentFinancialSeasonYear ??
      leagueData.currentFinancialSeasonYear ??
      existing.appliedThroughSeasonYear ??
      existing.appliedInflationThroughSeason ??
      options.currentFinancialSeasonYear ??
      getLeagueSeasonYear(leagueData),
    baseSeasonYear
  );
  const annualInflationRate = safeNumber(existing.annualInflationRate, DEFAULT_ANNUAL_INFLATION_RATE);
  const baseRules = buildBaseRulesFromLeague({ ...leagueData, financials: { ...existing, baseRules: existing.baseRules } });

  const next = {
    ...leagueData,
    financials: {
      version: LEAGUE_FINANCIALS_VERSION,
      ...existing,
      baseSeasonYear,
      annualInflationRate,
      baseRules,
      currentSeasonYear: currentFinancialSeasonYear,
      currentFinancialSeasonYear,
      appliedThroughSeasonYear: validSeasonYear(
        existing.appliedThroughSeasonYear ?? existing.appliedInflationThroughSeason ?? currentFinancialSeasonYear,
        currentFinancialSeasonYear
      ),
      history: existing.history && typeof existing.history === "object" ? existing.history : {},
    },
  };

  const rules = getLeagueFinancialRules(next, currentFinancialSeasonYear);
  const historyKey = String(currentFinancialSeasonYear);
  next.financials.history = {
    ...next.financials.history,
    [historyKey]: {
      ...(next.financials.history?.[historyKey] || {}),
      seasonYear: currentFinancialSeasonYear,
      inflationIndex: rules.inflationIndex,
      salaryCap: rules.salaryCap,
      luxuryTaxLine: rules.luxuryTaxLine,
      firstApron: rules.firstApron,
      secondApron: rules.secondApron,
      minimumSalary: rules.minimumSalary,
      minimumException: rules.minimumException,
      maxSalary: rules.maxSalary,
      midLevelException: rules.midLevelException,
      taxpayerMLE: rules.taxpayerMLE,
      roomException: rules.roomException,
      twoWaySalary: rules.twoWaySalary,
    },
  };

  return normalizeLeagueFinancialAliases(next, rules);
}

export function applyLeagueInflationForOffseason(leagueData = {}, targetSeasonYear = null) {
  if (!leagueData || typeof leagueData !== "object") return leagueData;

  const ensured = ensureLeagueFinancials(leagueData);
  const targetYear = validSeasonYear(targetSeasonYear ?? getLeagueSeasonYear(ensured) + 1);
  const financials = ensured.financials || {};
  const alreadyApplied = validSeasonYear(
    financials.appliedThroughSeasonYear ?? financials.appliedInflationThroughSeason ?? financials.currentSeasonYear,
    getLeagueSeasonYear(ensured)
  );

  if (alreadyApplied >= targetYear) {
    return normalizeLeagueFinancialAliases(ensured, getLeagueFinancialRules(ensured, targetYear));
  }

  const next = {
    ...ensured,
    currentFinancialSeasonYear: targetYear,
    financials: {
      ...financials,
      version: LEAGUE_FINANCIALS_VERSION,
      currentSeasonYear: targetYear,
      currentFinancialSeasonYear: targetYear,
      appliedThroughSeasonYear: targetYear,
      appliedInflationThroughSeason: targetYear,
      lastAppliedAt: new Date().toISOString(),
    },
  };

  const rules = getLeagueFinancialRules(next, targetYear);
  next.financials.history = {
    ...(financials.history || {}),
    [String(targetYear)]: {
      seasonYear: targetYear,
      inflationIndex: rules.inflationIndex,
      salaryCap: rules.salaryCap,
      luxuryTaxLine: rules.luxuryTaxLine,
      firstApron: rules.firstApron,
      secondApron: rules.secondApron,
      hardCap: rules.hardCap,
      minimumSalary: rules.minimumSalary,
      minimumException: rules.minimumException,
      veteranMinimum: rules.veteranMinimum,
      maxSalary: rules.maxSalary,
      midLevelException: rules.midLevelException,
      taxpayerMLE: rules.taxpayerMLE,
      roomException: rules.roomException,
      twoWaySalary: rules.twoWaySalary,
      appliedAt: next.financials.lastAppliedAt,
    },
  };

  return normalizeLeagueFinancialAliases(next, rules);
}

export function getRookieSalaryForPick(leagueData = {}, roundNum = 1, pickNum = 1, seasonYear = null) {
  const ensured = ensureLeagueFinancials(leagueData || {});
  const rules = getLeagueFinancialRules(ensured, seasonYear ?? getCurrentFinancialSeasonYear(ensured));
  const baseRules = ensured.financials?.baseRules || DEFAULT_BASE_FINANCIAL_RULES;
  const inflationIndex = Number(rules.inflationIndex || 1);
  const roundNumber = Number(roundNum || 1);
  const pickNumber = Math.max(1, Number(pickNum || 1));

  let baseSalary;
  if (roundNumber === 1) {
    baseSalary = Math.max(
      Number(baseRules.rookieFirstRoundFloor || DEFAULT_BASE_FINANCIAL_RULES.rookieFirstRoundFloor),
      Number(baseRules.rookiePick1Salary || DEFAULT_BASE_FINANCIAL_RULES.rookiePick1Salary) -
        (pickNumber - 1) * Number(baseRules.rookieFirstRoundDecline || DEFAULT_BASE_FINANCIAL_RULES.rookieFirstRoundDecline)
    );
  } else {
    const pickInRound = Math.max(1, pickNumber - 30);
    baseSalary = Math.max(
      Number(baseRules.rookieSecondRoundFloor || DEFAULT_BASE_FINANCIAL_RULES.rookieSecondRoundFloor),
      Number(baseRules.rookieSecondRoundBase || DEFAULT_BASE_FINANCIAL_RULES.rookieSecondRoundBase) -
        (pickInRound - 1) * Number(baseRules.rookieSecondRoundDecline || DEFAULT_BASE_FINANCIAL_RULES.rookieSecondRoundDecline)
    );
  }

  return roundMoney(baseSalary * inflationIndex, 1_000);
}

export function salaryToCapPercent(salary, leagueData = {}, seasonYear = null) {
  const rules = getLeagueFinancialRules(leagueData || {}, seasonYear);
  if (!rules.salaryCap) return 0;
  return Number(salary || 0) / Number(rules.salaryCap || 1);
}
